File: allteams.php
<?php
  /*
  This doesn't work so great.  It's supposed to total up the points allocated for each team's stat totals, and it sucks at it.




  */
  $starttime=time();
  //print $starttime;
  include('simple_html_dom.php');
  include('functions.php');
  include('db_pass.php');
  $db = mysql_connect("localhost", $mysql_user, $mysql_pass);
  mysql_select_db("matthear_fantasybaseball",$db);

  print_header();

  /*print "<HR><B>Warning</B>: The actual stats are correct (I think) but all the calculations of points and results are shit. (You can tell 'cause Kas and Brian aren't projected to run away with the league.)  Ignore them until I can, you know, fix them.  You'll know that happened when I remove this warning.<HR>";*/

  $statarr=array("R","HR","RBI","SB","BAVG","K","W","SV","ERA","WHIP");

  /*function sort_by ($a, $b) {
    return $a['R'] - $b['R'];
  }*/

  if (isset($_GET['season'])) {
    $season=$_GET['season'];
  } else {
    $season=date('Y');
  } 

  #get number of teams:
  $team_query="select ID,Owner from FantasyTeams where Season=$season order by Owner";
  #print $team_query."<P>\n";
  $team_result=mysql_query($team_query);
  $numteams=mysql_num_rows($team_result);

  #print "numteams: ".$numteams."<P>\n";
   
  foreach ($statarr as $statstr) {
    $max[$statstr]=0;
    $min[$statstr]=10000;
  }
    
  while ($team_row=mysql_fetch_array($team_result)) {
    $teamid=$team_row[ID];
    #print $teamid;
    $player_query="select Players.PosString as PosString,FantasyTeams.ID as TeamID, FantasyTeams.Owner as Owner,Players.Name as PlayerName,DraftResult.Pick as DraftPick,";
    foreach ($statarr as $statstr) {
      $player_query=$player_query."Players.$statstr as $statstr";
      if ($statstr!="WHIP") $player_query=$player_query.", ";
    }
    #$teamstat['ID']=$teamid;
    $teamstat[$teamid]['Owner']=$team_row['Owner'];
    print $player_row['Owner'];
    $player_query=$player_query." from DraftResult,Players,FantasyTeams where DraftResult.FantasyTeam=$teamid and DraftResult.Player=Players.ID and DraftResult.Season=$season and FantasyTeams.ID=DraftResult.FantasyTeam order by DraftResult.Pick desc";
    #print $player_query;
    $player_result=mysql_query($player_query);
    $teamstat[$teamid]['numplayers']=mysql_num_rows($player_result); 
    while ($player_row=mysql_fetch_array($player_result)) {
      foreach ($statarr as $statstr) {
        if ($statstr=="BAVG" and $player_row['BAVG']>0) {
	  $teamstat[$teamid]['Hits']+=($player_row['BAVG']*500);
          $teamstat[$teamid]['AtBats']+=500; 
        } else {
          if ($statstr=="ERA" and $player_row['ERA']>0) {
            if ($player_row['W'] < 10) { #consider him a reliever
              $teamstat[$teamid]['RA']+=$player_row['ERA']*60;
	    } else { #consider him a starter
              $teamstat[$teamid]['RA']+=$player_row['ERA']*150;
            }
          } else {
            if ($statstr=="WHIP" and $player_row['WHIP']>0) {
              if ($player_row['W'] < 10) { #consider him a reliever
                $teamstat[$teamid]['WHA']+=$player_row['WHIP']*60;
                $teamstat[$teamid]['Innings']+=60;
              } else { #consider him a starter
                $teamstat[$teamid]['WHA']+=$player_row['WHIP']*150;
                $teamstat[$teamid]['Innings']+=150;
              }
            } else {
              $teamstat[$teamid][$statstr]+=$player_row[$statstr];
            }
          }
        }
      }
    }
    $teamstat[$teamid]['BAVG']=round($teamstat[$teamid]['Hits']/$teamstat[$teamid]['AtBats'],3);
    $teamstat[$teamid]['ERA']=round($teamstat[$teamid]['RA']/$teamstat[$teamid]['Innings'],2);
    $teamstat[$teamid]['WHIP']=round($teamstat[$teamid]['WHA']/$teamstat[$teamid]['Innings'],2);
    foreach ($statarr as $statstr) {
      $average[$statstr]+=$teamstat[$teamid][$statstr];
      if ($max[$statstr]<$teamstat[$teamid][$statstr]) $max[$statstr]=$teamstat[$teamid][$statstr];
      if ($min[$statstr]>$teamstat[$teamid][$statstr]) $min[$statstr]=$teamstat[$teamid][$statstr];
    }
  }
  foreach ($statarr as $statstr) {
    $average[$statstr]=round($average[$statstr]/$numteams,3);
    $diff[$statstr]=$max[$statstr]-$min[$statstr];
    $statppts[$statstr]=$diff[$statstr]/($numteams-1);
    #print $statstr." max: ".$max[$statstr]."<BR>\n";
    #print $statstr." min: ".$min[$statstr]."<BR>\n";
    #print $statstr." diff: ".$diff[$statstr]."<BR>\n";
    #print $statstr." statppts: ".$statppts[$statstr]."<BR>\n";
  }

  foreach ($teamstat as $teamid=>$teamiter) {
    foreach ($statarr as $statstr) {
      if ($statstr == "ERA" or $statstr == "WHIP") {
        $teamstat[$teamid]['Pts']+=($max[$statstr]-$teamstat[$teamid][$statstr])/$statppts[$statstr]+1;
        $teampts[$teamid][$statstr]=($max[$statstr]-$teamstat[$teamid][$statstr])/$statppts[$statstr]+1;
      } else {
        $teamstat[$teamid]['Pts']+=(($teamstat[$teamid][$statstr]-$min[$statstr])/$statppts[$statstr])+1;
        $teampts[$teamid][$statstr]=($teamstat[$teamid][$statstr]-$min[$statstr])/$statppts[$statstr]+1;
      }
    }
  }

  #usort($teamstat, 'sort_by ');

  #print "<pre>";
  #var_dump($teamstat);
  #print "</pre>";

  #var_dump($teamstat[$teamid]);
  #print "<PRE>";
  ##var_dump($teamstat);
  #var_dump($teampts);
  #print "</PRE>";
  function sort_by_Pts($a, $b)
  {
    return $b['Pts'] - $a['Pts'];
  }

  #print "<HR><HR><P>\n";
  uasort($teamstat, 'sort_by_Pts');
  #print "<PRE>";
  #var_dump($teamstat);
  #var_dump($teampts);
  #print "</PRE>";

  print "<TABLE BORDER=1>\n";
  print "<TR>";
  print "<TD>Owner</TD>";
  print "<TD>Players<BR>(of 26)</TD>";
  foreach ($statarr as $statitem) {
    print "<TD>".$statitem."</TD><TD>pts</TD>"; 
  }
  print "<TD>Points</TD></TR>\n";
  foreach ($teamstat as $teamid=>$teamiter) {
    print "<TR>";
    #print "<TD>";
    #var_dump($teamiter);
    #print $teamid;
    #print "</TD>";
    print "<TD><A HREF=http://www.matthearn.com/fantasybaseball/team.php?teamid=$teamid>".$teamiter['Owner']."</A></TD>";
    print "<TD>".$teamiter['numplayers']."</TD>";
    foreach ($statarr as $statitem) {
      print "<TD>".$teamiter[$statitem]."</TD>"; 
      print "<TD>".round($teampts[$teamid][$statitem],1)."</TD>"; 
    }
    print "<TD>".round($teamiter['Pts'],2)."</TD>"; 
    print "</TR>\n";
  }
  print "<TR><TD>Average</TD>";
  foreach ($statarr as $statitem) {
    print "<TD></TD>";
    print "<TD>".$average[$statitem]."</TD>"; 
  }
  print "</TR>\n";
  print "</TABLE>\n";

  print "Processing time: ".(time()-$starttime)."s<P>\n";
?>

File: allteams.py
#!/ramdisk/bin/python

from subprocess import Popen, PIPE
from collections import OrderedDict

query="select * from FantasyTeams where (Season=2016)"
teamstable = Popen(["mysql","-N","-u","matthear_hearnmd","","-e",query,"matthear_fantasybaseball"], stdin=PIPE, stdout=PIPE, stderr=PIPE)
#output, err = teamstable.communicate(b"")
#print output


statmax = OrderedDict([('R', 0.0), ('HR', 0.0), ('RBI', 0.0), ('SB', 0.0), ('BAVG', 0.0), ('K', 0.0), ('W', 0.0), ('SV', 0.0), ('ERA', 0.0), ('WHIP', 0.0)])
statmin = OrderedDict([('R', 1000.0), ('HR', 1000.0), ('RBI', 1000.0), ('SB', 1000.0), ('BAVG', 1000.0), ('K', 1000.0), ('W', 1000.0), ('SV', 1000.0), ('ERA', 1000.0), ('WHIP', 1000.0)])

statdict = OrderedDict() 
ownerdict = OrderedDict() 

for teamline in iter(teamstable.stdout.readline,''):
  #print teamline
  teamid=teamline.split("	")[0]
  #print teamid + "	",
  teamname=teamline.split("	")[2] 
  ownerdict[teamid]=teamname
  #print teamid
  query="select * from DraftResult,Players where (DraftResult.Player=Players.ID and DraftResult.Season=2016 and DraftResult.FantasyTeam=" + str(teamid) + ") order by DraftResult.Pick;"
  p = Popen(["mysql","-N","-u","matthear_hearnmd","","-e",query,"matthear_fantasybaseball"], stdin=PIPE, stdout=PIPE, stderr=PIPE)

  #statdict.update(OrderedDict([((teamid,'R'),0.0),((teamid,'HR'),0.0),((teamid,'RBI'),0.0),((teamid,'SB'),0.0),((teamid,'BAVG'),0.0),((teamid,'K'),0.0),((teamid,'W'),0.0),((teamid,'SV'),0.0),((teamid,'ERA'),0.0),((teamid,'WHIP'),0.0)]))
  totdict=OrderedDict([('R',0.0),('HR',0.0),('RBI',0.0),('SB',0.0),('BAVG',0.0),('K',0.0),('W',0.0),('SV',0.0),('ERA',0.0),('WHIP',0.0)])
  #print "totdict:"
  #print totdict
  #statdict.update(OrderedDict([((teamid,'R'), 0.0), ((teamid,'HR'), 0.0), ((teamid,'RBI'): 0.0, (teamid,'SB'): 0.0, (teamid,'BAVG'): 0.0, (teamid,'K'): 0.0, (teamid,'W'): 0.0, (teamid,'SV'): 0.0, (teamid,'ERA'): 0.0, (teamid,'WHIP'): 0.0]))
  pcount = 0
  bcount = 0

  for line in iter(p.stdout.readline,''):
    #print line
    name=line.split("	")[8]
    pos=line.split("	")[10].split(",")[0]
    #print name + " " + pos
    if pos == "SP" or pos == "RP":
      #print "He's a pitcher, shitass"
      totdict['K']+=float(line.split("	")[18])
      totdict['W']+=float(line.split("	")[19])
      totdict['SV']+=float(line.split("	")[20])
      totdict['ERA']+=float(line.split("	")[21])
      totdict['WHIP']+=float(line.split("	")[22])
      pcount+=1
    else:
      #print "He's a batter, dumbnut"
      totdict['R']+=float(line.split("	")[13])
      totdict['HR']+=float(line.split("	")[14])
      totdict['RBI']+=float(line.split("	")[15])
      totdict['SB']+=float(line.split("	")[16])
      totdict['BAVG']+=float(line.split("	")[17])
      bcount+=1

  #print totdict

  for item in ["R","HR","RBI","SB","BAVG"]:
    statdict[teamid,item]=round(totdict[item]/bcount,3)
    #print str(statdict[teamid,item]) + "	",
    if statdict[teamid,item] > statmax[item]:
      #print "Changing max for " + item + " from " + str(statmax[item]) + " to " + str(statdict[teamid,item])
      statmax[item]=statdict[teamid,item]
    if statdict[teamid,item] < statmin[item]:
      statmin[item]=statdict[teamid,item]

  for item in ["K","W","SV","ERA","WHIP"]:
    statdict[teamid,item]=round(totdict[item]/pcount,3)
    #print str(statdict[teamid,item]) + "	",
    if statdict[teamid,item] > statmax[item]:
      #print "Changing max for " + item + " from " + str(statmax[item]) + " to " + str(statdict[teamid,item])
      statmax[item]=statdict[teamid,item]
    if statdict[teamid,item] < statmin[item]:
      statmin[item]=statdict[teamid,item]

  #print

#for statkey in statmax.keys():
  #print str(statkey) + " " + str(statmax[statkey]) + " " + str(statmin[statkey])

ptsdict=OrderedDict()

for teamkey in statdict.keys():
  #print teamkey
  teamid=teamkey[0]
  statstr=teamkey[1]
  ptspstat=float((statmax[statstr]-statmin[statstr])/11)
  #print "teamid: " + str(teamid)
  #print "stat: " + statstr
  #print "statmax: " + str(statmax[statstr])
  #print "statdict: " + str(statdict[(teamid,statstr)])
  #print "statmin: " + str(statmin[statstr])
  #print "ptspstat: " + str(ptspstat)
  if statstr=="ERA" or statstr=="WHIP":
    ptsdict[(teamid,statstr)]=12-((statdict[(teamid,statstr)]-statmin[statstr])/ptspstat)
  else:
    ptsdict[(teamid,statstr)]=((statdict[(teamid,statstr)]-statmin[statstr])/ptspstat)+1

#print ptsdict

oldid=0
sumpts=0.0

print "ID	Owner	R		HR		RBI		SB		BAVG		K		W		SV		ERA		WHIP		Total"
for teamkey in statdict.keys():
  teamid=teamkey[0]
  statstr=teamkey[1]
  if not(teamid==oldid):
    sumpts=0.0
    print ownerdict[teamid] + "	",
    oldid=teamid
    print str(teamid) + "	",
  print str(round(statdict[teamid,statstr],3)) + "	",
  print str(round(ptsdict[teamid,statstr],1)) + "	",
  sumpts+=ptsdict[teamid,statstr]
  if statstr is "WHIP":
    print str(round(sumpts,1)) 
#print
